File: rev_sum/rev_sum.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var head *ListNode = nil
	var result *ListNode = head
	carry := 0
	for l1 != nil || l2 != nil {
		sum := carry
		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}
		carry = 0

		if head == nil {
			head = new(ListNode)
			result = head
		} else {
			result.Next = new(ListNode)
			result = result.Next
		}
		result.Val = sum
		if sum >= 10 {
			result.Val -= 10
			carry++
		}
	}

	if carry > 0 {
		result.Next = new(ListNode)
		result.Next.Val = 1
	}
	return head
}

File: rev_sum/rev_sum.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* result = nullptr;
        ListNode*  result_tail = result;
        ListNode* l1_cur = l1, *l2_cur = l2;
        int carry = 0;
        while(l1_cur != nullptr || l2_cur != nullptr) {
            int sum = carry + (l1_cur != nullptr ? l1_cur->val : 0) + (l2_cur != nullptr ? l2_cur->val : 0);
            carry = 0;
            if(result_tail == nullptr) {
                result = result_tail = new ListNode(sum);
            } else {
                result->next = new ListNode(sum);
                result = result->next;
            }
            if(sum >= 10) {
                result->val -= 10;
                carry++;
            }
            if(l1_cur != nullptr)
                l1_cur = l1_cur->next;
            if(l2_cur != nullptr)
                l2_cur = l2_cur->next;
        }
        if(carry) {
            result->next = new ListNode(1);
        }
        return result_tail;
    }

};